### PR TITLE
Created an actual Message type and moved Plan out of federated

### PR DESCRIPTION
**Pull request by iamtrask**

### CODE REVIEW

File: syft/federated/__init__.py [hunk 1]
```diff
@@ -1,10 +1,5 @@
-from syft.federated.plan import Plan
 from syft.federated.train_config import TrainConfig
 from syft.federated import federated_client
 from syft.federated.federated_client import FederatedClient
 
-from syft.federated.plan import func2plan
-from syft.federated.plan import method2plan
-from syft.federated.plan import make_plan
-
```
Comment (robert-wagner) [inline syft/federated/__init__.py]: All needs to change as well

File: syft/serde/serde.py [hunk 2]
```diff
@@ -48,7 +48,7 @@
 from syft.workers import AbstractWorker
 from syft.workers import VirtualWorker
 
-from syft.federated import Plan
+from syft.msg import Plan, Message
```
Comment (robert-wagner) [inline syft/serde/serde.py]: Put these on separate lines

File: syft/workers/base.py [hunk 3]
```diff
@@ -85,6 +85,8 @@ def __init__(
         self.verbose = verbose
         self.auto_add = auto_add
         self.msg_history = list()
+        self.promises = {}
+        self.tensor2promise_id = {}
```
Comment (robert-wagner) [inline syft/workers/base.py]: What is this parameter for?

File: syft/msg/__init__.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+from .message import Message
```
Comment (robert-wagner) [inline syft/msg/__init__.py]: Use absolute imports please

File: syft/msg/message.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import syft as sy
+
+from syft.workers import AbstractWorker
+
+
+class Message:
```
Comment (robert-wagner) [inline syft/msg/message.py]: Should plans inherit from message? Also this class seems like it might be a security risk, what prevents me from creating an arbitrary message?

File: syft/workers/base.py [hunk 6]
```diff
@@ -217,8 +220,9 @@ def send_msg(self, msg_type: int, message: str, location: "BaseWorker") -> objec
         """
         if self.verbose:
             print(f"worker {self} sending {msg_type} {message} to {location}")
+
         # Step 0: combine type and message
-        message = (msg_type, message)
+        message = sy.msg.Message(msg_type, message)
```
Comment (robert-wagner) [inline syft/workers/base.py]: This would look much cleaner if you imported message. That being said, what is this point if this change, why not just keep it a tuple

File: syft/workers/base.py [hunk 7]
```diff
@@ -251,7 +255,10 @@ def recv_msg(self, bin_message: bin) -> bin:
             self.msg_history.append(bin_message)
 
         # Step 0: deserialize message
-        (msg_type, contents) = sy.serde.deserialize(bin_message, worker=self)
+        msg = sy.serde.deserialize(bin_message, worker=self)
```
Comment (robert-wagner) [inline syft/workers/base.py]: Same as above for my question about tuples

File: syft/workers/base.py [hunk 8]
```diff
@@ -843,6 +850,9 @@ def generate_triple(
         c_shared = c.share(*locations, field=field, crypto_provider=self).child
         return a_shared, b_shared, c_shared
 
+    def save_promise(self):
```
Comment (robert-wagner) [inline syft/workers/base.py]: Why does this method that does nothing exist?

File: syft/workers/base.py [hunk 9]
```diff
@@ -863,4 +873,6 @@ def create_message_execute_command(
         """
         if return_ids is None:
             return_ids = []
-        return tuple([codes.MSGTYPE.CMD, [[command_name, command_owner, args, kwargs], return_ids]])
+        return sy.Message(
```
Comment (robert-wagner) [inline syft/workers/base.py]: Again, what advantage does this give us over a tuple

File: test/workers/test_virtual.py [hunk 10]
```diff
@@ -93,7 +93,7 @@ def test_recv_msg():
     # Test 2: get tensor back from alice
 
     # Create message: Get tensor from alice
-    msg = (MSGTYPE.OBJ_REQ, obj.id)
+    msg = sy.msg.Message(MSGTYPE.OBJ_REQ, obj.id)
```
Comment (robert-wagner) [inline test/workers/test_virtual.py]: Why are there no other tests for messages?

File: syft/federated/__init__.py (hunk 1 above)
Comment (iamtrask) [reply syft/federated/__init__.py]: done

File: syft/serde/serde.py (hunk 2 above)
Comment (iamtrask) [reply syft/serde/serde.py]: Done

File: syft/workers/base.py (hunk 3 above)
Comment (iamtrask) [reply syft/workers/base.py]: oops - bit early for that - this was originally a bigger PR but i had to stop because it was already getting quite large (better to merge along the way)

File: syft/msg/__init__.py (hunk 4 above)
Comment (iamtrask) [reply syft/msg/__init__.py]: done

File: syft/msg/message.py (hunk 5 above)
Comment (iamtrask) [reply syft/msg/message.py]: Not any more of a security risk than it was before, messages are still whitelisted based on the msg_type.  This is literally just a class where the tuple used to be which carried messages. Starting small.

File: syft/__init__.py [hunk 11]
```diff
@@ -48,11 +48,14 @@
 from syft.grid import VirtualGrid
 
 # Import federate learning objects
-from syft.federated import Plan
```
Comment (midokura-silvia) [inline syft/__init__.py]: Could we limit this pull request to only contain the introduction of the message type?
It is not clear to me, why Plan should be in a message folder.

File: syft/workers/base.py (hunk 6 above)
Comment (iamtrask) [reply syft/workers/base.py]: It's a myriad of reasons, mostly organization. I'm working on some more advanced async/protocol/promise functionality and it's become quite annoying that messages aren't an actual type with actual to-strings which help with introspection.

File: syft/codes.py [hunk 12]
```diff
@@ -8,6 +8,7 @@ class MSGTYPE(object):
     GET_SHAPE = 7
     SEARCH = 8
     FORCE_OBJ_DEL = 9
+    PROMISE = 10
```
Comment (midokura-silvia) [inline syft/codes.py]: is this within the scope of this pull request?

File: syft/workers/base.py (hunk 6 above)
Comment (iamtrask) [reply syft/workers/base.py]: eventually Message objects will be passed to send_msg but I didn't want to do too much in one PR. This is an incremental PR for sure.

File: syft/workers/base.py (hunk 7 above)
Comment (iamtrask) [reply syft/workers/base.py]: same answer as above

File: syft/workers/base.py (hunk 8 above)
Comment (iamtrask) [reply syft/workers/base.py]: removed

File: syft/workers/base.py (hunk 9 above)
Comment (iamtrask) [reply syft/workers/base.py]: i'm working toward something which will give us more advantages - but i didn't want to make too many changes for one PR - so this is an incremental one.

File: test/workers/test_virtual.py (hunk 10 above)
Comment (iamtrask) [reply test/workers/test_virtual.py]: There is no new functionality for Message yet. I literally just replaced the tuple with a class with two attributes. Zillions of other tests already use it.

File: syft/workers/base.py [hunk 13]
```diff
@@ -843,6 +850,9 @@ def generate_triple(
         c_shared = c.share(*locations, field=field, crypto_provider=self).child
         return a_shared, b_shared, c_shared
 
+    def save_promise(self):
+        ""
```
[the remark this answers is not in the document]
Comment (iamtrask) [reply syft/workers/base.py]: removed it instead

File: syft/__init__.py (hunk 11 above)
Comment (iamtrask) [reply syft/__init__.py]: The message folder will contain all messages, groups of messages, and graphs of messages. Plan was in the federated folder which implied that it's only useful for Federated Learning.

File: syft/codes.py (hunk 12 above)
Comment (iamtrask) [reply syft/codes.py]: removed

File: test/workers/test_virtual.py (hunk 10 above)
Comment (midokura-silvia) [reply test/workers/test_virtual.py]: There might be tests that are using it, but all of them are integration tests (meaning, that they use something that uses the class Message). It would be good to have a basic unit test as well.

Comment (iamtrask) [reply test/workers/test_virtual.py]: fair - i'll make a test :)

Comment (iamtrask) [reply test/workers/test_virtual.py]: i'll test serde - which is new

File: syft/workers/base.py (hunk 6 above)
Comment (robert-wagner) [reply syft/workers/base.py]: Still would look cleaner importing message

Comment (iamtrask) [reply syft/workers/base.py]: Also we should be testing message construction - which we are not.

File: syft/workers/base.py (hunk 9 above)
Comment (robert-wagner) [reply syft/workers/base.py]: Still should import message though

File: syft/workers/base.py (hunk 6 above)
Comment (iamtrask) [reply syft/workers/base.py]: This will also allow us to add functionality inside the messages to make sure they're well formed. At present we only find out if they end up causing an error on the other side.

File: syft/__init__.py (hunk 11 above)
Comment (midokura-silvia) [reply syft/__init__.py]: I agree that Plan can be used for more than federated learning. But I am not sure that it should be in a message folder neither.

File: syft/workers/base.py (hunk 6 above)
Comment (iamtrask) [reply syft/workers/base.py]: Also this will make the code easier to write - at present it's totally undocumented what goes into a message (aka - our protocol is undocumented). A good first step here is setting out messages types and what information goes in each message.

Comment (iamtrask) [reply syft/workers/base.py]: This will also allow us to do smarter simplification/compression in the future

Comment (iamtrask) [reply syft/workers/base.py]: So yeah - lots of good reasons I think. :)

File: test/workers/test_virtual.py (hunk 10 above)
Comment (iamtrask) [reply test/workers/test_virtual.py]: groovy - added a test for message serde

File: syft/workers/base.py (hunk 9 above)
Comment (iamtrask) [reply syft/workers/base.py]: Done

File: syft/workers/base.py (hunk 6 above)
Comment (iamtrask) [reply syft/workers/base.py]: also - changed it to import Message directly

File: syft/msg/message.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+            ptr (Message): a Message
+        Returns:
+            tuple: a tuple holding the unique attributes of the message
+        Examples:
+            data = simplify(ptr)
+        """
+
+        return (ptr.msg_type, sy.serde._simplify(ptr.contents))
+
+        # a more general but slower/more verbose option
```
Comment (robert-wagner) [inline syft/msg/message.py]: Should this comment be here?

File: test/message/test_message.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
+import syft as sy
+import syft
+
+import torch as th
```
Comment (robert-wagner) [inline test/message/test_message.py]: Why are all these imports here?

File: syft/workers/base.py (hunk 6 above)
Comment (midokura-silvia) [reply syft/workers/base.py]: My prefered style is to use:
from sy import msg
and then use msg.Message(...)
That way you always have the source referenced. 
And additionally you follow the google style guide, see http://google.github.io/styleguide/pyguide.html, section 2.2

File: syft/__init__.py (hunk 11 above)
Comment (iamtrask) [reply syft/__init__.py]: a plan is a cache of messages - seemed pretty close. Other ideas for what to name the folder?

Comment (iamtrask) [reply syft/__init__.py]: maybe rename the folder "messaging"?

File: syft/workers/base.py (hunk 6 above)
Comment (iamtrask) [reply syft/workers/base.py]: lol - ok - will fix

Comment (iamtrask) [reply syft/workers/base.py]: fixed

File: syft/msg/message.py (hunk 14 above)
Comment (iamtrask) [reply syft/msg/message.py]: removed

File: test/message/test_message.py (hunk 15 above)
Comment (iamtrask) [reply test/message/test_message.py]: removed

File: test/message/test_plan.py [hunk 16]
```diff
@@ -420,9 +421,9 @@ def forward(self, x):
 
 def test_replace_worker_ids_two_strings(hook):
     plan = sy.Plan(id="0", owner=hook.local_worker, name="test_plan")
-    _replace_message_ids_orig = sy.federated.Plan._replace_message_ids
+    _replace_message_ids_orig = sy.msg.Plan._replace_message_ids
```
Comment (midokura-silvia) [inline test/message/test_plan.py]: Could you unify the way the import is done? (sy.Message vs sy.msg.Message vs sy.msg.Plan)
sy.msg.Plan => msg.Plan

File: test/message/test_message.py [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+
+
+def test_message_serde(hook):
+
+    x = sy.Message(0, [1,2,3])
+    x_bin = sy.serde.serialize(x)
+    y = sy.serde.deserialize(x_bin, sy.local_worker)
+
+    assert x.contents == y.contents
+    assert x.msg_type == y.msg_type
```
Comment (midokura-silvia) [inline test/message/test_message.py]: missing newling

File: syft/__init__.py (hunk 11 above)
Comment (midokura-silvia) [reply syft/__init__.py]: @mari-linhares what do you think about the naming of the folder and the location of Plan?

File: test/message/test_message.py (hunk 17 above)
Comment (iamtrask) [reply test/message/test_message.py]: done

File: test/message/test_plan.py (hunk 16 above)
Comment (iamtrask) [reply test/message/test_plan.py]: done

File: syft/__init__.py (hunk 11 above)
Comment (mari-linhares) [reply syft/__init__.py]: I agree that we should rename it eventually. Not sure if "message" is the best option but I don't have a good naming in mind neither.

Comment (iamtrask) [reply syft/__init__.py]: msg, cache,... something. 

It's relatively easy to change in the future with modern IDEs. But I think all are better than having it in federated.
